db/snapshot/backup_task.cc

```diff
@@ -118,8 +133,30 @@ future<> backup_task_impl::do_backup() {
     _backup_shard = this_shard_id();
     co_await _sharded_worker.start(std::ref(_snap_ctl.db()), _table_id, std::ref(*this));
 
+    gate abort_gate;
+    auto abort_sub = _as.subscribe([&] () noexcept {
+        if (auto gh = abort_gate.try_hold()) {
+            std::ignore = _sharded_worker.invoke_on_all([] (worker& m) {
+                m.abort();
+            }).finally([gh = std::move(gh)] {});
+        }
+    });
+
+    // Errors from `start_uploading` are retrieved by `done`
+    co_await _sharded_worker.invoke_on_all([] (worker& m) {
+        return m.start_uploading();
+    });
+
     try {
-        co_await uploads_worker();
+        co_await _sharded_worker.invoke_on_all([] (worker& m) {
+            return m.done();
```

Comment:
Why the pattern here is

```c++
    co_await sharded_worker.invoke_on_all(&worker::start);
    co_await sharded_worker.invoke_on_all(&worker::done);
```

isn't it enough if `worker::start()` just doesn't resolve until it's done, so that the caller goes with the single co_await invoke-on-all?

Reply:
It should be possible to return non-ready future from start_uploading (and rename it to e.g. `upload`). 
I think I did it this way since start_uploading was initially launched in the (synchronous) constructor.